src/pbab-mapping.ts

```diff
@@ -176,6 +188,123 @@ export function handleTransfer(event: Transfer): void {
     transfer.save();
   }
 }
+
+export function handleExternalAssetDependencyUpdated(
+  event: ExternalAssetDependencyUpdated
+): void {
+  let project = Project.load(
+    generateContractSpecificId(event.address, event.params._projectId)
+  );
+
+  if (!project) {
+    return;
+  }
+
+  const assetEntity = new ProjectExternalAssetDependency(
+    generateProjectExternalAssetDependencyId(
+      project.id,
+      event.params._index.toString()
+    )
+  );
+  assetEntity.cid = event.params._cid;
+  assetEntity.project = project.id;
+  assetEntity.index = event.params._index;
+  assetEntity.dependencyType =
+    FLEX_CONTRACT_EXTERNAL_ASSET_DEP_TYPES[event.params._dependencyType];
+  assetEntity.save();
+
+  project.externalAssetDependencyCount =
+    event.params._externalAssetDependencyCount;
+  project.updatedAt = event.block.timestamp;
+  project.save();
+}
+
+export function handleExternalAssetDependencyRemoved(
+  event: ExternalAssetDependencyRemoved
+): void {
+  let project = Project.load(
+    generateContractSpecificId(event.address, event.params._projectId)
+  );
+
+  if (!project) {
+    return;
+  }
+
+  let contract = GenArt721Core2EngineFlex.bind(event.address);
+  // Remove last asset entity on the project
+  const lastEntityIndex = project.externalAssetDependencyCount.minus(
+    BigInt.fromI32(1)
+  );
+  const entity = ProjectExternalAssetDependency.load(
+    generateProjectExternalAssetDependencyId(
+      project.id,
+      lastEntityIndex.toString()
+    )
+  );
+
+  if (entity) {
+    store.remove("ProjectExternalAssetDependency", entity.id);
+  }
+
+  // Refresh project external asset dependencies
+  for (
+    let index = BigInt.fromI32(0);
+    index < lastEntityIndex;
+    index = index.plus(BigInt.fromI32(1))
+  ) {
+    const assetEntity = new ProjectExternalAssetDependency(
+      generateProjectExternalAssetDependencyId(project.id, index.toString())
+    );
```

Comment:
I think all existing ProjectExternalAssetDependencies need to be removed from the store before creating the new ones here. This might appear to work on the surface, but I think there will be multiple entities at the same ID under the hood.
TBH, I'm fuzzy on how The Graph behaves when entity is created at the same ID, I think it gets weird, but maybe would just overwrite? Either way, if the last item was not the one removed, we would still have a stale item in the store, even if The Graph overwrites items. We had this type of issue when tx hash was our id for sales entities, and multiple token sales of a given token were happening in the same tx.

Reply:
good call on this 👍

Reply:
Can confirm, from testing, that the graph overwrites entities as one would expect. Tests also confirm that multiple entities at the same ID do not get left behind.

> Either way, if the last item was not the one removed, we would still have a stale item in the store, even if The Graph overwrites items. We had this type of issue when tx hash was our id for sales entities, and multiple token sales of a given token were happening in the same tx.

Because of the way the removal operation is written in the smart contract, it's ALWAYS the last item removed. I've added a comment above the handler to make sure this is very clear. I've also added to the removal event handler tests to make sure that there are no extraneous entities that get left and that the shifted data is where it should be

Reply:
ah, yes, thanks for helping me understand the overwrite behavior here! 🙏 

My original comment about "we would have a stale item regardless of overwrite behavior" is definitely incorrect - a "stale" item would only exist if same ID didn't overwrite. Thus, this LGTM and appreciate the extra test check to confirm this behavior 🔥